onnx/defs/tensor/defs.cc

```diff
@@ -108,8 +108,8 @@ ONNX_OPERATOR_SET_SCHEMA(
             OpSchema::all_tensor_types(),
             "Constrain input and output types to all tensor types.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
-          // Type inference
-          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          // Type inference, pass true to tell it is reshape 
+          propagateElemTypeFromInputToOutput(ctx, 0, 0, true);
```

Comment:
Why do we need this? The general assumption in type-inference is that input-types are known. What is the use-case / scenario where this does not hold?

Reply:
For [tests ](https://github.com/onnx/onnx/blob/cfab05a5f1cf33065a7a2ccc3df8020707e8998f/onnx/test/shape_inference_test.py#L42), these two inputs are set to TensorProto.UNDEFINED. I thought it should be right here because I have no idea what else tensor types we can set for them. Therefore, I chose to modify the checker part not to check the tensors type for reshape ops. 

BTW, actually there are many conflicts in tests after applying inference errors catching because the inference check is kind of general (every tensors need type definition) and there might be some exceptions. For example, ["Loop"](https://github.com/onnx/onnx/blob/cfab05a5f1cf33065a7a2ccc3df8020707e8998f/onnx/defs/controlflow/old.cc#L750). There are some unknown tensors like "M" or "cond" will trigger the inference error [here](https://github.com/onnx/onnx/blob/cfab05a5f1cf33065a7a2ccc3df8020707e8998f/onnx/defs/shape_inference.h#L165). 

My current solution is to make checker not check tensor types for these particular ops (reshape and Loop). In that case, I don't need to modify the inference graph and break something. Or maybe we could just let inference errors become warnings? Do you have any idea about how to modify it correctly?

Reply:
For the first part: I think it would be better to change TestShapeInference if possible, instead of changing the shape-inference code.

Reply:
For the loop case: where is the test that is failing? Again, even with a Loop, the expectation is that when we analyze a complete model, we should be able to infer the types inside a Loop from the outer context.

Reply:
> For the first part: I think it would be better to change TestShapeInference if possible, instead of changing the shape-inference code.

If so, I am thinking to use TensorProto.String instead of TensorProto.UNDEFINED.

Reply:
> For the loop case: where is the test that is failing? Again, even with a Loop, the expectation is that when we analyze a complete model, we should be able to infer the types inside a Loop from the outer context.

Test is [Here](https://github.com/pytorch/pytorch/blob/master/test/onnx/test_pytorch_onnx_caffe2.py#L2215). Tests involving forloop or while will all fail. I think the tests should be OK so I focus on modifying the checker part.  

Originally the checker will check input including "M" and "cond", but I don't know where to remove them from the graph. Do you know where to modify the input and output of graph (g_) [here](https://github.com/onnx/onnx/blob/cfab05a5f1cf33065a7a2ccc3df8020707e8998f/onnx/shape_inference/implementation.cc#L434)?  (In other words, which function calls the Loop in old.cc?) Not sure whether it is changeable. If yes, we might remove them in shape_inference because they look useless for shape inference.

Reply:
> > For the first part: I think it would be better to change TestShapeInference if possible, instead of changing the shape-inference code.
> 
> If so, I am thinking to use TensorProto.String instead of TensorProto.UNDEFINED in shape_inference_test.py.

NVM. Just found a better way to do this. Should use the type from the seed_value.

Reply:
It fails [here](https://github.com/onnx/onnx/blob/cfab05a5f1cf33065a7a2ccc3df8020707e8998f/onnx/defs/shape_inference.h#L165). I suspect the reason is the inference graph contains "M" and "cond" and both of them are TensorProto.UNDEFINED. And the output of Loop node is TensorProto.UNDEFINED, too. 

Yes, I can successfully create such model from PyTorch and it looks normal on Netron. 

My understanding is shape inference works normal for the inference part. Except for checking the tensors: it needs each tensor has defined tensor type. In that case, there are some inputs and outputs (like a condition) for "Loop" do not meet this requirement. 

To avoid checking this kind of tensors, I need to modify the input and output of the graph [here](https://github.com/onnx/onnx/blob/cfab05a5f1cf33065a7a2ccc3df8020707e8998f/onnx/shape_inference/implementation.cc#L434) as mentioned. However, I am still looking for where is the code for calling the Loop Node (construct the inference graph) or how can I modify the inference graph without breacking the logic.

Reply:
The first question we should answer is: is there a problem in the model (this implies that the problem is with the exporter that created the model) or not (in this case the bug is in the type inference code).

Reply:
Agree. I just checked the code on the Pytorch test  and found that the loop model can pass the onnx.checker.check_model(model) first but fail on the onnx.shape_inference.infer_shapes(model). I will review the Pytorch exporter part as well. Thank you for the reminder.

Reply:
@gramalingam Thank you for the previous instructions for solving the input issue. 

Another problem is about checking the output. The Loop node checker fails [here](https://github.com/onnx/onnx/blob/e094e101da8cd2a10d6bf5f5ab692b50943a1470/onnx/defs/controlflow/old.cc#L588). There are two output for Loop node and both of them are `TypeProto::ValueCase::VALUE_NOT_SET` [here](https://github.com/onnx/onnx/blob/e094e101da8cd2a10d6bf5f5ab692b50943a1470/onnx/shape_inference/implementation.cc#L470). It seems that there is a bug for the graph output not updated as well, but the output of Loop model looks normal (has valid type) by Netron. 

Since we cannot solve this problem in the same way as the input (not providing inferred type), what would you suggest to solve this? (Maybe skip the output checker?)

Reply:
I think the problem is here: https://github.com/onnx/onnx/blob/e094e101da8cd2a10d6bf5f5ab692b50943a1470/onnx/shape_inference/implementation.cc#L267 ... when we infer the type of X, we need to check if X is a graph output. If it is a graph output, we should update the type of the output, instead of adding a new valueinfo. So, we need another "else if" branch to handle this.